### PR TITLE
tp: _interval_interesect supports instant events

### PR DESCRIPTION
**Pull request by aMayzner**
_interval_intersect macro will now return instants if there are instant results. Removed the dur > 0 check and added tests. If the interval (ts, dur) is (0, 10) it will return instant when it's inside it, (5, 0), but the ends are excluded, so there is no overlap with (10, 2).

### CODE REVIEW

File: src/trace_processor/perfetto_sql/intrinsics/functions/type_builders.cc [hunk 1]
```diff
@@ -330,13 +330,21 @@ struct IntervalTreeIntervalsAgg
           ctx, "Interval intersect requires intervals to be sorted by ts.");
       return;
     }
-    agg_ctx.last_interval_start = interval.start;
     int64_t dur = sqlite::value::Int64(argv[2]);
-    if (dur < 1) {
-      sqlite::result::Error(
-          ctx, "Interval intersect only works on intervals with dur > 0");
+
+    // Ignore instants. We shouldn't error out on them, as they might be valid
```
Comment (LalitMaganti) [inline src/trace_processor/perfetto_sql/intrinsics/functions/type_builders.cc]: IMO instants should be passed through as-is as that's what most people would want/expect.

File: test/trace_processor/diff_tests/stdlib/intervals/intersect_tests.py [hunk 2]
```diff
@@ -328,30 +328,74 @@ def test_single_interval(self):
...
+        FROM _interval_intersect_single!(1, 6, A)
         ORDER BY ts;
         """,
         out=Csv("""
         "id","ts","dur"
         0,1,1
-        1,3,2
+        1,3,0
         2,6,1
         """))
```
Comment (LalitMaganti) [inline test/trace_processor/diff_tests/stdlib/intervals/intersect_tests.py:401]: Please add some better tests for multi-interval intersection with instants. Also test edge cases (what happens on instant at end of interval/start of interval etc.)

File: test/trace_processor/diff_tests/stdlib/intervals/intersect_tests.py [hunk 3]
```diff
@@ -328,33 +328,274 @@ def test_single_interval(self):
...
+    return DiffTestBlueprint(
+        trace=TextProto(""),
+        #      0 1 2 3 4 5 6 7
+        # A:   _ - - - - - - _
+        # B:   . _ . . _ . _ .
+        # res: _ _ . . _ . _ _
+        query="""
+        INCLUDE PERFETTO MODULE intervals.intersect;
+
+        CREATE PERFETTO TABLE B_instants AS
```
Comment (LalitMaganti) [inline test/trace_processor/diff_tests/stdlib/intervals/intersect_tests.py:577]: I believe instants on the edges of intervals should be passed through.

File: test/trace_processor/diff_tests/stdlib/intervals/intersect_tests.py [hunk 4]
```diff
@@ -328,33 +328,274 @@ def test_single_interval(self):
...
+        "id","ts","dur"
+        """))
+
+  def test_instants(self):
+    return DiffTestBlueprint(
+        trace=TextProto(""),
+        #      0 1 2 3 4 5 6 7
+        # A:   _ - - - - - - _
+        # B:   - - _ _ _ _ - -
+        # res: _ - _ - - _ - _
```
Comment (LalitMaganti) [inline test/trace_processor/diff_tests/stdlib/intervals/intersect_tests.py]: This ascii diagram does not make sense.

File: src/trace_processor/containers/interval_tree.h [hunk 5]
```diff
@@ -39,6 +39,23 @@ struct Interval {
   Id id;
 };
 
+inline bool IsOverlapping(bool query_is_instant,
+                          Ts s,
+                          Ts e,
+                          const Interval& i) {
+  bool interval_is_instant = (i.start == i.end);
+  if (query_is_instant && interval_is_instant) {
```
Comment (LalitMaganti) [inline src/trace_processor/containers/interval_tree.h:47]: I believe instants at both the start and end of the interval should be passed through.
